larpandoracontent/LArObjects/LArThreeDSlidingFitResult.cc

```diff
@@ -224,12 +224,20 @@ TrackState ThreeDSlidingFitResult::GetPrimaryAxis(const CartesianPointVector *co
     LArPcaHelper::RunPca(*pPointVector, centroid, eigenValues, eigenVecs);
 
     float minProjection(std::numeric_limits<float>::max());
-    const CartesianVector &fitDirection(eigenVecs.at(0));
+    CartesianVector fitDirection(eigenVecs.at(0));
+
+    // By convention, the primary axis has a positive z-component.
+    // However, downstream algorithms should be independent of this convention.
+    if (fitDirection.GetZ() < 0.f)
+        fitDirection *= -1.f;
 
     for (const CartesianVector &coordinate : *pPointVector)
         minProjection = std::min(minProjection, fitDirection.GetDotProduct(coordinate - centroid));
 
-    return TrackState(centroid + (fitDirection * minProjection), fitDirection);
+    // Define layers based on centroid rather than individual extremal hits
+    const float fitProjection(layerPitch * std::floor(minProjection / layerPitch));
```

Comment:
Is there a way that minProjection might still be std::numeric_limits<float>::max() at this point?

Yes, but only if all the fitDirection.GetDotProduct(coordinate - centroid) results on line 235 also return float max. Technically possible, but vanishingly unlikely in a real use case.